compiler/resolution/functionResolution.cpp

```diff
@@ -1628,6 +1628,166 @@ computeGenericSubs(SymbolMap &subs,
+        if (firstArg->var->type == dtMethodToken) {
+          // If this fails, ensure that isVarargSizeExpr still works correctly.
+          INT_ASSERT(parent->get(2) == se);
+          return true;
+        }
+  return false;
+}
+
+//
+// Does 'ast' contain use(s) of the vararg tuple 'formal'
+// that require 'formal' as a whole tuple?
+//
+// Keep in sync with substituteVarargTupleRefs().
+//
+static bool needVarArgTupleAsWhole(BlockStmt* ast, int numArgs,
+                                   ArgSymbol* formal)
+{
+  std::vector<SymExpr*> symExprs;
+  collectSymExprs(ast, symExprs);
+
+  for_vector(SymExpr, se, symExprs) {
+    if (se->var == formal) {
+      if (CallExpr* parent = toCallExpr(se->parentExpr)) {
+        SET_LINENO(parent);
+        if (parent->isPrimitive(PRIM_TUPLE_EXPAND)) {
+          // (...formal) -- does not require
+          continue;
+        }
+        if (isVarargAccessIndex(se, parent, numArgs)) {
+          // formal(i) -- does not require
+          continue;
+        }
+        if (isVarargSizeExpr(se, parent)) {
+          // formal.size -- does not require
+          continue;
+        }
+      }
+      // Something else ==> may require.
+      return true;
+    }
+  }
+
+  // The 'formal' is used only in "does not require" cases.
+  return false;
+}
+
+//
+// Replace, in 'ast', uses of the vararg tuple 'formal'
+// with references to individual formals.
+//
+// Keep in sync with needVarArgTupleAsWhole().
+//
+void substituteVarargTupleRefs(BlockStmt* ast, int numArgs, ArgSymbol* formal,
+                               std::vector<ArgSymbol*>& varargFormals)
+{
+  std::vector<SymExpr*> symExprs;
+  collectSymExprs(ast, symExprs);
+  SymbolMap tempReps;
+
+  for_vector(SymExpr, se, symExprs) {
+    if (se->var == formal) {
+      if (CallExpr* parent = toCallExpr(se->parentExpr)) {
+        SET_LINENO(parent);
+
+        if (parent->isPrimitive(PRIM_TUPLE_EXPAND)) {
+          // Replace 'parent' with a sequence of individual args.
+          for_vector(ArgSymbol, arg, varargFormals)
+            parent->insertBefore(new SymExpr(arg));
+          parent->remove();
+          continue;
+        }
+
+        if (int idxNum = isVarargAccessIndex(se, parent, numArgs)) {
+          INT_ASSERT(1 <= idxNum && idxNum <= numArgs);
+          ArgSymbol* replFormal = varargFormals[idxNum-1];
+          replaceVarargAccess(parent, se, replFormal, tempReps);
+          continue;
+        }
+
+        if (isVarargSizeExpr(se, parent)) {
+          parent->replace(new SymExpr(new_IntSymbol(numArgs)));
+          continue;
+        }
+
+      }
+      // Cases not "continue"-ed above should have been ruled out
+      // by needVarArgTupleAsWhole().
+      INT_ASSERT(false);
+    } else if (Symbol* replmt = tempReps.get(se->var)) {
+      SET_LINENO(se);
+      se->replace(new SymExpr(replmt));
+    }
+  }
+}
+
+
 /** Common code for multiple paths through expandVarArgs.
```

Comment:
On May 11, 2016, at 1:58 PM, Vassily Litvinov <notifications@github.com<mailto:notifications@github.com>> wrote:


In compiler/resolution/functionResolution.cpp<https://github.com/chapel-lang/chapel/pull/3838#discussion_r62925637>:

> +          parent->replace(new SymExpr(new_IntSymbol(numArgs)));
> +          continue;
> +        }
> +
> +      }
> +      // Cases not "continue"-ed above should have been ruled out
> +      // by needVarArgTupleAsWhole().
> +      INT_ASSERT(false);
> +    } else if (Symbol* replmt = tempReps.get(se->var)) {
> +      SET_LINENO(se);
> +      se->replace(new SymExpr(replmt));
> +    }
> +  }
> +}
> +
> +
>  /** Common code for multiple paths through expandVarArgs.


I suggest doing such refactoring as a separate PR, as it is orthogonal to the change I am making here. OK with you?


I don’t think that it’s orthogonal.  I think this work has (a) exposed an existing challenge
in the code base (b) exacerbated the problem.  That’s pretty common of course.


I agree that it’s great to do factorings in small pieces when practical and I
would have attempted that too.   To be clear the path would be

1) Start to do some development of a new component.

2) Realize the integration is going to be challenging.

3) Do the appropriate non-functional refactorings to enable the
original goal.

4) Apply the new component to the refactored code.


The alternative order tends to lead to an increasing tangle of code with
comments that say “come back and clean this up”.  A bit like our conversation
on generic function instatiation.


What are you proposing?

Mike



—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub<https://github.com/chapel-lang/chapel/pull/3838/files/20b76cba1018c724ee2d739174a99cfcfbb22a17#r62925637>